File: src/testing/oracle-runner.ts
import type { AgentCLI } from '../runtime.js';
import type { OracleCommandSpec, ScenarioSpec } from './scenario.js';

export interface OracleStepTrace {
  command: string;
  output: string;
  body: string;
  exitCode: number;
  expectedExitCode: number;
}

export interface OracleTrace {
  scenarioId: string;
  steps: OracleStepTrace[];
  finalOutput: string;
  finalBody: string;
  finalExitCode: number;
  totalDurationMs: number;
}

interface PresentedOutput {
  body: string;
  exitCode: number;
}

export async function runOracle(runtime: AgentCLI, scenario: ScenarioSpec): Promise<OracleTrace> {
  if (scenario.oracle === undefined || scenario.oracle.length === 0) {
    throw new Error(`scenario ${scenario.id} does not define an oracle`);
  }

  const steps: OracleStepTrace[] = [];
  let finalOutput = '';
  let finalBody = '';
  let finalExitCode = 0;
  const started = performance.now();

  for (const oracleStep of scenario.oracle) {
    const normalizedStep = normalizeOracleStep(oracleStep);
    const output = await runtime.run(normalizedStep.command);
    const presented = parsePresentedOutput(output);

    steps.push({
      command: normalizedStep.command,
      output,
      body: presented.body,
      exitCode: presented.exitCode,
      expectedExitCode: normalizedStep.expectedExitCode,
    });
    finalOutput = output;
    finalBody = presented.body;
    finalExitCode = presented.exitCode;
  }

  return {
    scenarioId: scenario.id,
    steps,
    finalOutput,
    finalBody,
    finalExitCode,
    totalDurationMs: Math.max(0, performance.now() - started),
  };
}

function normalizeOracleStep(step: string | OracleCommandSpec): Required<OracleCommandSpec> {
  if (typeof step === 'string') {
    return {
      command: step,
      expectedExitCode: 0,
    };
  }

  return {
    command: step.command,
    expectedExitCode: step.expectedExitCode ?? 0,
  };
}

function parsePresentedOutput(output: string): PresentedOutput {
  const footerMatch = output.match(/\[exit:(\d+) \| [^\]]+\]\s*$/);
  if (!footerMatch || footerMatch.index === undefined) {
    throw new Error(`oracle output is missing an exit footer:\n${output}`);
  }

  const body = output.slice(0, footerMatch.index).replace(/\n+$/, '');
  return {
    body,
    exitCode: Number(footerMatch[1]),
  };
}

File: python/src/onetool/testing/oracle_runner.py
from __future__ import annotations

import time
from dataclasses import dataclass

from ..runtime import AgentCLI, RunExecution
from .scenario import OracleCommandSpec, ScenarioSpec


@dataclass(frozen=True, slots=True)
class OracleStepTrace:
    command: str
    output: str
    body: str
    exit_code: int
    expected_exit_code: int
    execution: RunExecution


@dataclass(frozen=True, slots=True)
class OracleTrace:
    scenario_id: str
    steps: list[OracleStepTrace]
    final_output: str
    final_body: str
    final_exit_code: int
    total_duration_ms: float


async def run_oracle(runtime: AgentCLI, scenario: ScenarioSpec) -> OracleTrace:
    if not scenario.oracle:
        raise ValueError(f"scenario {scenario.id} does not define an oracle")

    steps: list[OracleStepTrace] = []
    final_output = ""
    final_body = ""
    final_exit_code = 0
    started = time.perf_counter()

    for raw_step in scenario.oracle:
        step = _normalize_oracle_step(raw_step)
        execution = await runtime.run_detailed(step.command)
        steps.append(
            OracleStepTrace(
                command=step.command,
                output=execution.presentation.text,
                body=execution.presentation.body,
                exit_code=execution.exit_code,
                expected_exit_code=step.expected_exit_code or 0,
                execution=execution,
            )
        )
        final_output = execution.presentation.text
        final_body = execution.presentation.body
        final_exit_code = execution.exit_code

    return OracleTrace(
        scenario_id=scenario.id,
        steps=steps,
        final_output=final_output,
        final_body=final_body,
        final_exit_code=final_exit_code,
        total_duration_ms=max(0.0, (time.perf_counter() - started) * 1000.0),
    )


def _normalize_oracle_step(step: str | OracleCommandSpec) -> OracleCommandSpec:
    if isinstance(step, str):
        return OracleCommandSpec(command=step, expected_exit_code=0)
    return OracleCommandSpec(
        command=step.command,
        expected_exit_code=0 if step.expected_exit_code is None else step.expected_exit_code,
    )
